Network/HTTP2/TLS/Client.hs

```diff
@@ -61,61 +67,96 @@ runTLS
     open = openTCP (settingsAddrInfoFlags settings) serverName port
-    params = getClientParams settings serverName alpn
+
+    -- TLS client parameters
+    params :: ClientParams
+    params =
+        getClientParams
+            settings
+            ( fromMaybe (H2Client.authority cliconf) $
+                settingsServerNameOverride settings
+            )
+            port
+            alpn
 
 -- | Running an HTTP\/2 client over TLS (over TCP).
-run :: Settings -> HostName -> PortNumber -> Client a -> IO a
-run settings serverName port client =
-    runTLS settings serverName port "h2" $ \ctx mysa peersa ->
-        run' settings "https" serverName (sendTLS ctx) (recvTLS ctx) mysa peersa client
+runWithConfig :: ClientConfig -> Settings -> HostName -> PortNumber -> Client a -> IO a
+runWithConfig cliconf settings serverName port client =
+    runTLSWithConfig cliconf settings serverName port "h2" $ \ctx mysa peersa ->
+        run' cliconf' (sendTLS ctx) (recvTLS ctx) mysa peersa client
+  where
+    cliconf' :: ClientConfig
+    cliconf' = cliconf{H2Client.scheme = "https"}
 
 -- | Running an HTTP\/2 client over TCP.
-runH2C :: Settings -> HostName -> PortNumber -> Client a -> IO a
-runH2C settings serverName port client =
+runH2CWithConfig :: ClientConfig -> HostName -> PortNumber -> Client a -> IO a
+runH2CWithConfig cliconf serverName port client =
     E.bracket open close $ \sock -> do
         mysa <- getSocketName sock
         peersa <- getPeerName sock
         recv <- mkRecvTCP Server.defaultSettings sock
-        run' settings "http" serverName (sendTCP sock) recv mysa peersa client
+        run' cliconf' (sendTCP sock) recv mysa peersa client
   where
     open = openTCP (settingsAddrInfoFlags defaultSettings) serverName port
 
+    cliconf' :: ClientConfig
+    cliconf' = cliconf{H2Client.scheme = "http"}
+
 run'
-    :: Settings
-    -> ByteString
-    -> HostName
+    :: ClientConfig
     -> (ByteString -> IO ())
     -> IO ByteString
     -> SockAddr
     -> SockAddr
     -> Client a
     -> IO a
-run' Settings{..} schm serverName send recv mysa peersa client =
+run' cliconf send recv mysa peersa client =
     E.bracket
         (allocConfigForClient send recv mysa peersa)
         freeConfigForClient
         (\conf -> H2Client.run cliconf conf client)
-  where
-    cliconf =
-        defaultClientConfig
-            { scheme = schm
-            , authority = C8.pack serverName
-            , cacheLimit = settingsCacheLimit
-            , connectionWindowSize = settingsConnectionWindowSize
-            , settings =
-                (settings defaultClientConfig)
-                    { initialWindowSize = settingsStreamWindowSize
-                    , maxConcurrentStreams = Just settingsConcurrentStreams
-                    }
-            }
+
+defaultClientConfig :: Settings -> ClientConfig
+defaultClientConfig Settings{..} =
+    H2Client.defaultClientConfig
+        { H2Client.scheme = "https"
```

Comment:
`defaultClientConfig` should also take `HostName` to override `H2Client.authority`.
If not, the backward compatibility is not maintained.

Reply:
Ah, you're quite right. Sorry for missing this. I have pushed another commit with a fix. I think this is now ready for merging, apart from perhaps [your comment above regarding your DNS over HTTP2/ test](https://github.com/kazu-yamamoto/http2-tls/pull/6#issuecomment-1865353335)..? Though I fail to see how this PR affects decoding :thinking: